mmit()

parismita opened the pull request:
A function to learn a MMIT model in R

Commit:
Add files via upload

Commit:
Update bestsplit.R

Commit:
Update bestsplit.R

Commit:
Add files via upload

Commit:
Update growtree.R

Commit:
Update growtree.R

Commit:
Update growtree.R

Commit:
Add files via upload

Commit:
Update bestsplit.R

Commit:
Update bestsplit.R

Commit:
Update growtree.R

Commit:
Update mmit.R

Commit:
Update growtree.R

Commit:
Update mmit.R

Commit:
Update mmit.R

Commit:
Update growtree.R

Commit:
Update growtree.R

Commit:
Update mmit.R

Commit:
Update growtree.R

Commit:
Add files via upload

aldro61 commented:
Thanks for the PR! I will have a look at the code. We will need working to get Travis to work before we can merge. Looks like it's not finding some required R packages.

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+
+    #1st and last index of duplicate elem
+    first_idx <- order(feat)[!duplicated(feat)]
+    last_idx <- c(order(feat)[!duplicated(feat)]-1,length(feat))
```
Coding standards: please use spaces after operators and commas.

E.g.:
c(order(feat)[!duplicated(feat)]-1,length(feat)) --> c(order(feat)[!duplicated(feat)] - 1, length(feat))

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,70 @@
+  
+
+  #if sample per node is less than minimum, we stop tree growth
+  if(length(target.mat[,1])<=min_sample) return(partynode(id = id, info = pred[[side]]))
```
Coding standards: spaces around <= please. Sorry for being picky about this :)

Commit:
Update test-mmit.R

Commit:
Update DESCRIPTION

Commit:
Update mmit.R

Commit:
Update growtree.R

Commit:
Update bestsplit.R

Commit:
Update DESCRIPTION

Commit:
Update test-mmit.R

Commit:
Update DESCRIPTION

Commit:
Update NAMESPACE

Commit:
Update NAMESPACE

parismita commented:
Testthat is failing here, but it was working in my pc, I solved the library issue, so what else can it be?

aldro61 commented:
It is not testthat that is failing, but travis. You should look at the travis log to see which error messages you find.

Le 24 mai 2018 04:38 -0400, parismita <notifications@github.com>, a écrit :
> Testthat is failing here, but it was working in my pc, I solved the library issue, so what else can it be?
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

aldro61 commented:
I’ll be reviewing the code later today or tomorrow, so I might be able to find what’s wrong.

parismita commented:
i looked into the log, testthat is failing for 5 cases, line where out$leftpref = 0.5

Commit:
Update test-mmit.R

the error in testthat is when I am not explicitly compiling the mmit, growtree and bestsplit, 
If those files are sourced explicitly, the error vanishes

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+  best_split$row <- NULL
+
+  #loop for every feature
+  for (index in 2:(length(feature.mat[1,]))){
```
Why do you start at index 2?

parismita replied:
the 1st index is the target...in mmit (line 6,7) I defined the 1st column to be the target

aldro61 replied:
Ok. It makes sense then. But it seems like an odd choice, since you also take target.mat as an argument. Why did you choose to have the 1st column be the target? Wouldn't it be more logic to have only features in feature.mat?

parismita replied:
True...but didn't think of that while making it...it seems redundant to pass target as well as add it in dataframe....shall I change it?

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+
+    #1st and last index of duplicate elem
+    first_idx <- order(feat)[!duplicated(feat)]
+    last_idx <- c(order(feat)[!duplicated(feat)]-1, length(feat))
```
You can reuse first_idx here instead of re-sorting

parismita replied:
ok

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+    last_idx <- last_idx[-1]
+
+    #if no unique elements
+    if(length(last_idx) == 0){
```
How is this possible? Each feature must have at least one value.

parismita replied:
if there is only [1,1,1,1] type data then unique elem =1, and i am deleting the 1st elem so to rule out such cases this was done

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+    feat <- z[sorted]
+    z <- target.mat
+    tar <- z[sorted,]
+    rev_tar <- z[rev(sorted),]   #reverse order
```
You don't need z.

```
rev_tar <- tar[rev(sorted), ]
tar <- tar[sorted, ]
```

parismita replied:
ok

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+
+    #sorted
+    sorted <- order(feat)
+    z <- feat
```
No need to create a new var here. See comment below.

parismita replied:
ok

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+    rightpred <- compute_optimal_costs(rev_tar, margin, loss)
+
+    #unique and removing cases where all examples are in one leaf
+    leftpred <- leftpred[last_idx,]
```
I didn't review the slicing and inverting that is done below in details. I'm trusting that you tested and compared with the Python implementation. Seems ok at first glance.

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    #compute cost, prediction
+    leftpred <- compute_optimal_costs(tar, margin, loss)
```
Maybe renaming to leftleaf and rightleaf would make things a bit clearer.

parismita replied:
ok

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+    rightpred <- rightpred[-1,]
+
+    #removing NA cases and summing both orders
+    split_cost <- na.omit(leftpred$cost + rightpred$cost)
```
Shouldn't be any NA values here. Did you encounter some? I don't find any equivalence for this block of code in the Python implementation.

parismita replied:
I encountered it once, but i guess it was because of my mistake, i'll remove it and check again if I am encountering it or not

parismita replied:
yeah its not required

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+    #removing NA cases and summing both orders
+    split_cost <- na.omit(leftpred$cost + rightpred$cost)
+
+    #if no split possible
```
Same comment as above. If I'm wrong, can you point me to the corresponding line in the Python code?

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,64 @@
+      next
+    }
+
+    if(min(split_cost) < best_split$cost){
```
This is not exactly correct. See here: https://github.com/aldro61/mmit/blob/master/mmit/learning.py#L181.

We want to compare the summed cost of the leaves resulting from each split. If no split can give a total cost lower than the cost of *the current leaf*, then it's not worth it to split and we don't split.

In other words, we only split if we find a split that decreases the cost of the current node. In your case, you are comparing to Inf, so you will always end up splitting.

aldro61 replied:
Don't forget, the node that we are trying to split is a leaf (until we split it), so it is also associated with a cost.

parismita replied:
if no split is better than splitting then what do we return? best_split$cost and other values as 0?

aldro61 replied:
I think you should return null in this case. Isn't this why you have the following line in growtree.R?

```
if (is.null(sp)) return(partynode(id = id, info = pred[[side]]))
```

In the python implementation, we simply return None as a marker that no split that decreased the current leaf's cost was found.

parismita replied:
done

Commit:
Update bestsplit.R

Commit:
Update bestsplit.R

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,72 @@
+
+  #splitting into kids
+  kidids <- kidids_split(sp, data = feature.mat) #list of right left kids
+  #print(length(kidids))
```
Don't forget to delete commented prints before we merge :)

parismita replied:
done

aldro61 commented on Rpackage/tests/testthat/test-mmit.R:
```diff
@@ -0,0 +1,65 @@
+
+#linear hinge, margin = 0, test best split
+out <- bestsplit(target.mat, feature.mat, margin=0.0, loss="hinge")
+test_that("margin=0.0 yields cost 0", {
```
You should rename the tests

parismita replied:
done

Commit:
Update bestsplit.R

Commit:
Update growtree.R

Commit:
Update mmit.R

Commit:
Update growtree.R

Commit:
Update mmit.R

Commit:
Update and rename test-mmit.R to test-tree.R

Commit:
Update mmit.R

Commit:
Update bestsplit.R

Commit:
Update test-tree.R

Commit:
Update growtree.R

Commit:
Update bestsplit.R

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,74 @@
+    last_idx <- last_idx[-1]
+
+    #if no unique elements
+    if(length(last_idx) == 0){
```
Change to one

parismita replied:
ok

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,74 @@
+    #summing both orders
+    split_cost <- leftleaf$cost + rightleaf$cost
+
+    #if no split possible
```
This should never happen if you have at least 2 values for a feat

parismita replied:
its giving this warning without that included

Warning messages:
1: In min(split_cost) : no non-missing arguments to min; returning Inf

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,74 @@
+  if(is.null(pred)){
+    return(best_split)
+  }
+  if(best_split$cost>=pred$cost){
```
You can include this check at line 53

aldro61 replied:
Or just leave it like that, its fine.

parismita replied:
ok

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,68 @@
+      myid <- id
+    }
+    depth <- depth+1
+    kids[[kidid]] <- growtree(tar[[kidid]], kid[[kidid]], pred=splt, side = 4+kidid, depth = depth,
```
Add comment for the 4+ and spaces :D

parismita replied:
done

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,68 @@
+growtree <- structure(function(target.mat, feature.mat, depth=0, maxdepth = Inf,
```
Ask Torsten for a code review for that part. Maybe he can make it more optimal.

aldro61 commented on Rpackage/R/mmit.R:
```diff
@@ -0,0 +1,23 @@
+mmit <- structure(function(formula, target.mat, feature.mat, weights = NULL,
```
Hey @tdhock! Do you think its better to receive one dataframe with all variables and then use a formula to extract targets and features, or to receive features and targets in separate matrices? I see that in your penalty learning code, you use the second option. Would like to have your opinion. Thanks!

aldro61 replied:
It seems like for most datasets (e.g., Iris dataset), the targets and features are in a single matrix. Is this common in R?

aldro61 commented on Rpackage/tests/testthat/test-tree.R:
```diff
@@ -0,0 +1,65 @@
+#linear hinge, margin = 0, test mmit
+out <- mmit(d~., target.mat, feature.mat)
+p <- predict(out)
+test_that("margin=0.0 yields cost 0", {
```
Change names to python test names

Commit:
Update bestsplit.R

Commit:
Update growtree.R

Commit:
Update test-tree.R

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,71 @@
+  if(length(kids) <= 1) return(partynode(id = id, info = pred[[side]]))
+
+  #creating list of data into
+  for(i in 1:length(feature.mat[,1])){
```
Torsten says that this should be changed to subsetting instead of creating a copy of the data. That is, we treat the data as constant and always pass it to the splitting function, while using one subset vector per child, that tells us which examples are in which child. 

It's basically the same idea as what you did here, but you avoid creating a copy of the data.

parismita replied:
yeah! working on it

Commit:
Update bestsplit.R

Commit:
Update growtree.R

Commit:
Update mmit.R

Commit:
Update test-tree.R

Commit:
Update test-tree.R

Commit:
Update bestsplit.R

Commit:
Update growtree.R

Commit:
Update growtree.R

Commit:
Update bestsplit.R

Commit:
Update mmit.R

Commit:
Update test-tree.R

Commit:
Update test-tree.R

Commit:
Update DESCRIPTION

Commit:
Add files via upload

aldro61 commented:
@parismita Let me know when you're ready for the final code review + merge

parismita commented:
yeah! I am ready

Commit:
Update mmit.Rd

Commit:
Update mmit.R

aldro61 commented on Rpackage/R/mmit.R:
```diff
@@ -0,0 +1,65 @@
+                           id = 1L, min_sample = 1) {
+  
+  ### arranging such that response is not present in feature.mat.
+  feature.mat <- model.frame(data = feature.mat)
```
What does this do? Do we still need it now that we don't use formulas?

parismita replied:
Yeah if we use model.response we need to define it as model.frame else it gives error

Commit:
Update mmit.R

Commit:
Update mmit.R

Commit:
Update growtree.R

Commit:
Update mmit.R

tdhock commented:
hi @parismita I noticed that you may be using the web interface to make changes to the code... is that true? Since the goal of GSOC is to help you learn free/ open source coding, I would recommend that you try editing the code with a real text editor (emacs or vim) or IDE (Rstudio) and then use the command line git to commit and push.

parismita commented:
yeah using web interface but coding in rstudio, i'll try git on next commit sure.

aldro61 commented:
@parismita @tdhock Good point! Yes, it will be better if you use git because you'll be able to assign informative names to your commits instead of "Update [...]".

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,72 @@
+                               margin = 0.0, loss="hinge", id = 1L, min_sample = 1,
+                               pred=NULL, side = NULL, weights = NULL) {
+  
+  pid <- NULL
```
Can you change pid to something a bit more informative? It's really not clear that it's the node info.

parismita replied:
done

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,72 @@
+    colnames(pid) <- c("prediction", "cost")
+  }
+
+  if(side == 10){
```
Why 10? We would need a comment to explain what is going on at that line.

parismita replied:
done

parismita committed:
comment about size = 10,
pid changed to node_info_print

Commit:
Delete .gitignore

Commit:
Delete mmit.Rproj

Commit:
Create .gitignore

aldro61 commented:
I recommend that you add a .gitignore file in the RPackage directory to avoid commiting unwanted files.

https://github.com/github/gitignore/blob/master/R.gitignore

parismita commented:
ok

parismita committed:
update gitignore

aldro61 commented on Rpackage/R/bestsplit.R:
```diff
@@ -0,0 +1,82 @@
+  }
+  
+  ### check if node cost is less than splitting cost
+  if(is.null(pred)){
```
What does this condition capture? Is it only true when we're in the root of the tree (i.e., no previous split)? If so, note that sometimes you will find that there exists no split that decreases the cost of the root. The decision to split should always be based on the cost of the current node (even for the root).

parismita replied:
yeah i'll include root nodes cost

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,73 @@
+  }
+  
+  ### passing side = 10 as arguement if node is root.
+  if(side == 10){
```
Do you have the guarantee that no other node will have side == 10? For instance, below you have the line `side = 4 + kidid`. Can there be any conflict? If not, why?

parismita replied:
nope, as pred has 6 list items, kidid is {1,2} as left and right, and side is {5,6}. so we can keep anything other than [1,6]. 

only conflict can be if the splitting is not into left right but into 4-5 different brances, like it happens in classification trees.
But this is regression tree so i guess there shouldnt be any conflict

parismita replied:
if kidid>2 then it might cause problem and side can only be null if its root so maybe i'll remove side = 10

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,73 @@
+  kids <- vector(mode = "list", length = max(kidids, na.rm = TRUE))
+  
+  ### if no left right kids, stop tree growth
+  if(length(kids) <= 1) return(partynode(id = id, info = node_info_print))
```
How can there be no left right kids if you split the node at line 40? `kidids <- kidids_split(sp, data = feature.mat) #list of right left kids`. I believe that if the split did not lead to two leaves, your code would have returned before that.

parismita replied:
if there is no splitting, case taken into account in line 36, i'll remove this

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,73 @@
+    } else {
+      myid <- id
+    }
+    depth <- depth+1
```
This looks like it causes a bug. You increment the depth at each kid. Shouldn't you increment only once? Both kids are at the same depth.

parismita replied:
yeah! i'll fix this

aldro61 replied:
Good! When you fix bugs like this, you should add a unit test to confirm that it works and avoid reintroducing the bug later.

parismita replied:
ok

parismita replied:
i think the increment in depth is fine as its recursive function and when left terminal is reached lets say it has depth d, growtree returns value to kids[[]] and in that loop depth is d-1  so doing +1 will make it correct depth.

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,73 @@
+    w[kidids != kidid] <- 0
+    
+    ### get next node id
+    if (kidid > 1) {
```
It's not clear to me what this does. Can you elaborate a little bit?

parismita replied:
kidid>1 means kidid =2, that is right child

Here we are assigning id to left and right child.
left = id+1 and right = max(node id of left kid)+1

eg if tree is like 
        1
  2                11
3   4          12 13
  5     8
6 7   9 10
and id =1
then leftid = 2, right id = last left node(10) +1 = 11

aldro61 commented on Rpackage/R/growtree.R:
```diff
@@ -0,0 +1,73 @@
+    #side = 4+kidid as pred[[5]] is left prediction and pred[[6]] right
+    kids[[kidid]] <- growtree(target.mat, feature.mat, pred = splt, side = 4 + kidid, depth = depth,
+                              maxdepth = maxdepth, margin = margin, loss = loss, 
+                              id = as.integer(myid + 1),min_sample = min_sample, weights = w)
```
Here also, the `id = as.integer(myid + 1)` is a bit obscur to me. Why do you add 1?

parismita replied:
explained in prev comment

aldro61 commented on Rpackage/R/mmit.R:
```diff
@@ -0,0 +1,66 @@
+#' @export
+mmit <- structure(function(target.mat, feature.mat, weights = NULL,
+                           depth=0, maxdepth = Inf, margin=0.0, loss="hinge",
+                           id = 1L, min_sample = 1) {
```
What is this `id` argument? It's not in the documentation and I don't think it's required.

aldro61 replied:
Some think for `weights`. There is not such argument in the python implementation.

parismita replied:
weights is just the name of array passed which determines which data to keep

parismita replied:
id is required for creation of partynode

aldro61 replied:
Yes, but these should not be specified by the user of the mmit function. They can be initialized to some value internally.

parismita replied:
shall i create a variable id inside mmit function and remove it from arguements?

parismita replied:
same with depth =0

aldro61 commented on Rpackage/R/mmit.R:
```diff
@@ -0,0 +1,66 @@
+  feature.mat <- model.frame(data = feature.mat)
+  
+  ### assigning all weights as 1 in the beginning. weights determine which data is to be considered in which node.
+  if (is.null(weights)) weights <- rep(1L, nrow(feature.mat))
```
Just initialize weights to a vector of ones. No need for more verifications. I don't see why weights should be user specified.

parismita replied:
yeah....will do that

parismita committed:
included root cost in bestsplit
removed split check as left right kid exists always in growtree
removed weight, id,depth from arguement of mmit

parismita committed:
removed side=10

parismita committed:
update comments

parismita committed:
depth outside loop

Commit:
change data package name to neuroblastoma

Commit:
include penaltyLearning library for testing

Commit:
Update NAMESPACE

parismita committed:
install penaltyLearning

parismita committed:
change test

parismita committed:
spaces

Commit:
Update .gitignore

aldro61 committed:
Augment depth in the recursive call

aldro61 committed:
Code style

aldro61 commented:
All good! Merging. Congrats on your first PR :)
